include/ginkgo/core/solver/batch_solver_base.hpp

```diff
@@ -0,0 +1,389 @@
+     * Get the maximum number of iterations set on the solver.
+     *
+     * @return  Maximum number of iterations.
+     */
+    int get_max_iterations() const { return max_iterations_; }
+
+    /**
+     * Set the maximum number of iterations for the solver to use,
+     * independent of the factory that created it.
+     *
+     * @param max_iterations  The maximum number of iterations for the solver.
+     */
+    void reset_max_iterations(int max_iterations)
+    {
+        if (max_iterations < 0) {
+            GKO_INVALID_STATE("Max iterations cannot be negative!");
+        }
+        max_iterations_ = max_iterations;
+    }
+
+    /**
+     * Get the tolerance type.
+     *
+     * @return  The tolerance type.
+     */
+    ::gko::batch::stop::tolerance_type get_tolerance_type() const
+    {
+        return tol_type_;
+    }
+
+    /**
+     * Set the type of tolerance check to use inside the solver
+     *
+     * @param tol_type  The tolerance type.
+     */
+    void reset_tolerance_type(::gko::batch::stop::tolerance_type tol_type)
+    {
+        if (tol_type == ::gko::batch::stop::tolerance_type::absolute ||
+            tol_type == ::gko::batch::stop::tolerance_type::relative) {
+            tol_type_ = tol_type;
+        } else {
+            GKO_INVALID_STATE("Invalid tolerance type specified!");
+        }
+    }
+
+protected:
+    BatchSolver() {}
+
+    BatchSolver(std::shared_ptr<const BatchLinOp> system_matrix,
+                std::shared_ptr<const BatchLinOp> gen_preconditioner,
+                const double res_tol, const int max_iterations,
+                const ::gko::batch::stop::tolerance_type tol_type)
+        : system_matrix_{std::move(system_matrix)},
+          preconditioner_{std::move(gen_preconditioner)},
+          residual_tol_{res_tol},
+          max_iterations_{max_iterations},
+          tol_type_{tol_type},
+          workspace_{}
+    {}
+
+    std::shared_ptr<const BatchLinOp> system_matrix_{};
+    std::shared_ptr<const BatchLinOp> preconditioner_{};
+    double residual_tol_{};
+    int max_iterations_{};
+    ::gko::batch::stop::tolerance_type tol_type_{};
+    mutable array<unsigned char> workspace_{};
+};
+
+
+/**
+ * The parameter type shared between all preconditioned iterative solvers,
+ * excluding the parameters available in iterative_solver_factory_parameters.
+ * @see GKO_CREATE_FACTORY_PARAMETERS
+ */
+struct preconditioned_iterative_solver_factory_parameters {
+    /**
+     * The preconditioner to be used by the iterative solver. By default, no
+     * preconditioner is used.
+     */
+    std::shared_ptr<const BatchLinOpFactory> preconditioner{nullptr};
+
+    /**
+     * Already generated preconditioner. If one is provided, the factory
+     * `preconditioner` will be ignored.
+     */
+    std::shared_ptr<const BatchLinOp> generated_preconditioner{nullptr};
+};
+
+
+template <typename Parameters, typename Factory>
+struct enable_preconditioned_iterative_solver_factory_parameters
```

Comment:
polymophism. I start thinking the class naming rule makes them inconsistent...

Reply:
Yes, but here I followed the same structure as in `solver_base.hpp`

Reply:
actually, if we follow the polymophism not just the inhertence, the base class without virtual is not polymophism

Reply:
I think subtyping is a type of polymorphism. So, I think this technically a polymorphic class. But I think it is more important to keep the same naming for the core solvers and the batch solvers. So, if we change one, we should change the other as well. I dont have a particular preference.